plugins/node/opentelemetry-instrumentation-undici/src/undici.ts

```diff
@@ -0,0 +1,486 @@
+  trace,
+  ValueType,
+} from '@opentelemetry/api';
+
+import { VERSION } from './version';
+
+import {
+  ListenerRecord,
+  RequestHeadersMessage,
+  RequestMessage,
+  RequestTrailersMessage,
+  ResponseHeadersMessage,
+} from './internal-types';
+import { UndiciInstrumentationConfig, UndiciRequest } from './types';
+import { SemanticAttributes } from './enums/SemanticAttributes';
+import {
+  hrTime,
+  hrTimeDuration,
+  hrTimeToMilliseconds,
+} from '@opentelemetry/core';
+
+interface InstrumentationRecord {
+  span: Span;
+  attributes: Attributes;
+  startTime: HrTime;
+}
+
+// A combination of https://github.com/elastic/apm-agent-nodejs and
+// https://github.com/gadget-inc/opentelemetry-instrumentations/blob/main/packages/opentelemetry-instrumentation-undici/src/index.ts
+export class UndiciInstrumentation extends InstrumentationBase {
+  // Keep ref to avoid https://github.com/nodejs/node/issues/42170 bug and for
+  // unsubscribing.
+  private _channelSubs!: Array<ListenerRecord>;
+  private _recordFromReq = new WeakMap<UndiciRequest, InstrumentationRecord>();
+
+  private _httpClientDurationHistogram!: Histogram;
+  constructor(config?: UndiciInstrumentationConfig) {
+    super('@opentelemetry/instrumentation-undici', VERSION, config);
+    this.setConfig(config);
+  }
+
+  // No need to instrument files/modules
+  protected override init() {
+    return undefined;
+  }
+
+  override disable(): void {
+    if (!this._config.enabled) {
+      return;
+    }
+
+    this._channelSubs.forEach(sub => sub.channel.unsubscribe(sub.onMessage));
+    this._channelSubs.length = 0;
+    this._config.enabled = false;
+  }
+
+  override enable(): void {
+    if (this._config.enabled) {
+      return;
+    }
+    this._config.enabled = true;
+
+    // This method is called by the `InstrumentationAbstract` constructor before
+    // ours is called. So we need to ensure the property is initalized
+    this._channelSubs = this._channelSubs || [];
+    this.subscribeToChannel(
+      'undici:request:create',
+      this.onRequestCreated.bind(this)
+    );
+    this.subscribeToChannel(
+      'undici:client:sendHeaders',
+      this.onRequestHeaders.bind(this)
+    );
+    this.subscribeToChannel(
+      'undici:request:headers',
+      this.onResponseHeaders.bind(this)
+    );
+    this.subscribeToChannel('undici:request:trailers', this.onDone.bind(this));
+    this.subscribeToChannel('undici:request:error', this.onError.bind(this));
+  }
+
+  override setConfig(config?: UndiciInstrumentationConfig): void {
+    super.setConfig(config);
+
+    if (config?.enabled) {
+      this.enable();
+    } else {
+      this.disable();
+    }
+  }
+
+  protected override _updateMetricInstruments() {
+    this._httpClientDurationHistogram = this.meter.createHistogram(
+      'http.client.request.duration',
+      {
+        description: 'Measures the duration of outbound HTTP requests.',
+        unit: 's',
+        valueType: ValueType.DOUBLE,
+      }
```

Comment:
Should we add a boundary advice here to avoid having the default boundaries?
see boundaries from semconv: https://opentelemetry.io/docs/specs/semconv/http/http-metrics/#http-client

Reply:
I think you're right. I will add it